JGit auth fail

Hi

I'm getting the following exception when trying to install the scalatra-sbt
template:

``` sh
$ g8 scalatra/scalatra-sbt
org.eclipse.jgit.api.errors.JGitInternalException: Exception caught during execution of fetch command
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:138)
        at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:175)
        at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:121)
        at giter8.Apply$class.clone(apply.scala:73)
        at giter8.Giter8.clone(giter8.scala:3)
        at giter8.Apply$class.inspect(apply.scala:33)
        at giter8.Giter8.inspect(giter8.scala:3)
        at giter8.Apply$class.inspect(apply.scala:55)
        at giter8.Giter8.inspect(giter8.scala:3)
        at giter8.Giter8$$anonfun$2.apply(giter8.scala:19)
        at giter8.Giter8$$anonfun$2.apply(giter8.scala:18)
        at scala.Option.map(Option.scala:133)
        at giter8.Giter8.run(giter8.scala:18)
        at giter8.Giter8.run(giter8.scala:10)
        at giter8.Giter8.run(giter8.scala:3)
        at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:57)
        at xsbt.boot.Launch$.withContextLoader(Launch.scala:77)
        at xsbt.boot.Launch$.run(Launch.scala:57)
        at xsbt.boot.Launch$$anonfun$explicit$1.apply(Launch.scala:45)
        at xsbt.boot.Launch$.launch(Launch.scala:65)
        at xsbt.boot.Launch$.apply(Launch.scala:16)
        at xsbt.boot.Boot$.runImpl(Boot.scala:32)
        at xsbt.boot.Boot$.main(Boot.scala:21)
        at xsbt.boot.Boot.main(Boot.scala)
Caused by: org.eclipse.jgit.errors.TransportException: git@github.com:scalatra/scalatra-sbt.g8.git: Auth fail
        at org.eclipse.jgit.transport.JschConfigSessionFactory.getSession(JschConfigSessionFactory.java:141)
        at org.eclipse.jgit.transport.SshTransport.getSession(SshTransport.java:121)
        at org.eclipse.jgit.transport.TransportGitSsh$SshFetchConnection.<init>(TransportGitSsh.java:248)
        at org.eclipse.jgit.transport.TransportGitSsh.openFetch(TransportGitSsh.java:147)
        at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:127)
        at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:113)
        at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1062)
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:129)
        ... 23 more
Caused by: com.jcraft.jsch.JSchException: Auth fail
        at com.jcraft.jsch.Session.connect(Session.java:461)
        at org.eclipse.jgit.transport.JschConfigSessionFactory.getSession(JschConfigSessionFactory.java:116)
        ... 30 more
```

JGit is failing to use ssh to clone the repository, and I guess its because my
key is protected by gpg-agent(which offers normal ssh-agent emulation). I also
tried to use a full URL with no success:

``` sh
$ g8 https://github.com/scalatra/scalatra-sbt.g8.git
scala.MatchError: https://github.com/scalatra/scalatra-sbt (of class java.lang.String)
        at giter8.Apply$class.inspect(apply.scala:24)
        at giter8.Giter8.inspect(giter8.scala:3)
        at giter8.Giter8$$anonfun$2.apply(giter8.scala:19)
        at giter8.Giter8$$anonfun$2.apply(giter8.scala:18)
        at scala.Option.map(Option.scala:133)
        at giter8.Giter8.run(giter8.scala:18)
        at giter8.Giter8.run(giter8.scala:10)
        at giter8.Giter8.run(giter8.scala:3)
        at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:57)
        at xsbt.boot.Launch$.withContextLoader(Launch.scala:77)
        at xsbt.boot.Launch$.run(Launch.scala:57)
        at xsbt.boot.Launch$$anonfun$explicit$1.apply(Launch.scala:45)
        at xsbt.boot.Launch$.launch(Launch.scala:65)
        at xsbt.boot.Launch$.apply(Launch.scala:16)
        at xsbt.boot.Boot$.runImpl(Boot.scala:32)
        at xsbt.boot.Boot$.main(Boot.scala:21)
        at xsbt.boot.Boot.main(Boot.scala)
Error during sbt execution: scala.MatchError: https://github.com/scalatra/scalatra-sbt (of class java.lang.String)
```

Any ideas on how I might work around this?

+1, using the new `sbt new` and GHE repo, I get:

```
Caused by: com.jcraft.jsch.JSchException: Auth fail
        at com.jcraft.jsch.Session.connect(Session.java:512)
        at org.eclipse.jgit.transport.JschConfigSessionFactory.getSession(JschConfigSessionFactory.java:116)
        at org.eclipse.jgit.transport.SshTransport.getSession(SshTransport.java:136)
        at org.eclipse.jgit.transport.TransportGitSsh$SshFetchConnection.<init>(TransportGitSsh.java:262)
        at org.eclipse.jgit.transport.TransportGitSsh.openFetch(TransportGitSsh.java:161)
        at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:136)
        at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:122)
        at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1138)
        at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:130)
        at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:193)
        at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:133)
        at giter8.JgitHelper$.clone(JgitHelper.scala:114)
        at giter8.JgitHelper$.run(JgitHelper.scala:65)
        at giter8.Giter8$$anonfun$2.apply(giter8.scala:34)
        at giter8.Giter8$$anonfun$2.apply(giter8.scala:33)
        at scala.Option.map(Option.scala:145)
        at giter8.Giter8.run(giter8.scala:33)
        at sbtgiter8resolver.Giter8TemplateResolver.run(Giter8TemplateResolver.scala:31)
```

--- a/cli-git/src/main/scala/GitInteractor.scala
+++ b/cli-git/src/main/scala/GitInteractor.scala
@@ -21,7 +21,7 @@ import java.io.File
 
 import giter8.GitInteractor.TransportError
 import org.eclipse.jgit.api.errors.TransportException
-import org.eclipse.jgit.transport.{CredentialsProvider, SshSessionFactory, SshTransport}
+import org.eclipse.jgit.transport.{CredentialsProvider, SshTransport}
 import org.eclipse.jgit.api.{Git => JGit}
 
 import scala.util.{Failure, Success, Try}
@@ -51,7 +51,7 @@ class JGitInteractor(knownHosts: Option[String]) extends GitInteractor {
       .setCredentialsProvider(ConsoleCredentialsProvider)
       .setTransportConfigCallback({
         case sshTransport: SshTransport => sshTransport.setSshSessionFactory(new SshAgentSessionFactory(knownHosts))
-        case x                          => x
+        case x                          => ()
       })
       .call()
       .close()

--- a/cli-git/src/main/scala/SshAgentSessionFactory.scala
+++ b/cli-git/src/main/scala/SshAgentSessionFactory.scala
@@ -1,39 +1,30 @@
 package giter8
 
-import java.nio.file.{Files, Paths}
+import org.eclipse.jgit.transport.sshd.SshdSessionFactory
+import java.{util => ju}
+import java.io.File
+import java.nio.file.{Path, Files, Paths}
+import scala.collection.JavaConverters._
 
-import com.jcraft.jsch.agentproxy.{ConnectorFactory, RemoteIdentityRepository}
-import com.jcraft.jsch.{JSch, Session}
-import org.eclipse.jgit.transport.{JschConfigSessionFactory, OpenSshConfig}
-import org.eclipse.jgit.util.FS
+class SshAgentSessionFactory(knownHosts: Option[String]) extends SshdSessionFactory() {
 
-class SshAgentSessionFactory(knownHosts: Option[String]) extends JschConfigSessionFactory() {
+  private val home        = Option(System.getProperty("user.home")).getOrElse("")
+  private val programData = Option(System.getenv("PROGRAMDATA")).getOrElse("")
 
-  private val Home        = Option(System.getProperty("user.home")).getOrElse("")
-  private val ProgramData = Option(System.getenv("PROGRAMDATA")).getOrElse("")
-
-  private val KnownHostsPaths: List[String] = List(
-    s"$Home/.ssh/known_hosts",
+  private val defaultKnownHosts: List[String] = List(
+    s"$home/.ssh/known_hosts",
+    s"$home/.ssh/known_hosts2",
     "/etc/ssh/ssh_known_hosts",
-    s"$ProgramData/ssh/ssh_known_hosts"
+    s"$programData/ssh/ssh_known_hosts"
   )
 
-  override protected def configure(host: OpenSshConfig.Host, session: Session): Unit = {}
-
-  override protected def createDefaultJSch(fs: FS): JSch = {
-    val jsch = new JSch()
-    knownHostsWithDefault.foreach(jsch.setKnownHosts)
-    jsch.setIdentityRepository(identityRepository)
-    jsch
+  override def getDefaultKnownHostsFiles(sshDir: File): ju.List[Path] = {
+    knownHosts
+      .map(_ :: Nil)
+      .getOrElse(defaultKnownHosts)
+      .map(Paths.get(_))
+      .filter(Files.exists(_))
+      .asJava
   }
 
-  private def knownHostsWithDefault = {
-    knownHosts.orElse(KnownHostsPaths.find(path => Files.exists(Paths.get(path))))
-  }
-
-  private def identityRepository = {
-    val factory   = ConnectorFactory.getDefault
-    val connector = factory.createConnector
-    new RemoteIdentityRepository(connector)
-  }
 }

